### DISCUSSION TITLE
Create Regex Attachment Handler

### DISCUSSION THREAD

**Discussion by JJWest12** [nextgenhealthcare/connect#6059] [Feature/Enhancement Ideas]
I am trying to write up a regex attachment handler that removes the final segment from an inbound HL7V2 message. The final segment is always a custom segment with the header "PLY".

My attachment handler regex is PLY.* with a MIME type of text/plain. When running this I can see that the attachment is made and on the end of the raw message on the source, however I get this error below:

TypeError: The content of elements must consist of well-formed character data or markup.

All datatypes on the channel are HL7V2. What am I missing here?

**Comment by pacmano1** [comment]
sample message.  you know better :). 

And an hl7 message with a header?  What does that mean?  You mean a final segment name of "PLY"?

**Reply by JJWest12** [reply]
Sample message is below. I starred out a whole bunch of stuff for PHI purposes. Essentially I want to remove the PLY segment at the end and make it an attachment


MSH|^~\&||****||***|20240104073426|****|MDM^T02^DOC|************************************************|2.5.1
EVN|T02|20240104073426||UPDATE_NOTE|****^********^******^^^^^^****^^^^^****
PID|||**********||****^* ***********^^^^^*||********|M||White|** *** **^***** **^**^*****^US^P^^******|^^******||(***)***-****^P^H^^^***^******|(***)***-***^P^W^^^***^*******|ENG|MARRIED|******||***-**-****|||NOT HISPANIC|||||No|||N
PV1||INPATIENT|**^****^***-*^**^^^^^^^*****||||*****^*******^*******^^^^^^****^^^^****||******^****^*******^*^^^^^****^^^^****|Hospitalist|||||||*****^**^****^^^^^^****^^^^****||*********|||||||||||||||||HOME||||||||**************|**************||****.**|||********
OBR|1||*************|||*************||^Progress Notes|||||||||||||||||||||F||||||||||
OBX|1|ED|******^Progress Notes|1|SEE ATTACHMENT||||||F|||
PLY|*************|application/pdf||lafkss9edfkfknshaye783sjkdkfmfshjkamsjdhfyyya777

**Reply by JJWest12** [reply]
and by "header" I meant the value all the way to the left on each line lol. I don't know that the correct term for that is

**Reply by kirbykn2** [reply]
PLY is the segment name.  When referring to PLY, MSH, PID, etc I just refer to them as the segment.

"The final segment is a custom segment called "PLY"."

**Reply by pacmano1** [reply]
I suspect your ask is to add the base64 PDF as an attachment, correct?    Meaning  from PLY-4, correct?    Then remove the PLY segment?

**Reply by jonbartels** [reply]
> The content of elements must consist of well-formed character data or markup.

and

> My attachment handler regex is PLY.* 

I bet what is happening is that your message is being turned into this:
```
MSH|^~&|||||20240104073426||MDM^T02^DOC||2.5.1
EVN|T02|20240104073426||UPDATE_NOTE|^^^^^^^^^^^^^
PID|||||^ ***********^^^^^|||M||White| *** ^ ^^^US^P^^|^^||()-^P^H^^^^|()-^P^W^^^^|ENG|MARRIED|||--|||NOT HISPANIC|||||No|||N
PV1||INPATIENT|**^^-^^^^^^^^||||^^^^^^^^^^^^||^^^^^^^^^^^^|Hospitalist|||||||***^^^^^^^^^^^^|||||||||||||||||||HOME|||||||||||****.|||*
OBR|1|||||||^Progress Notes|||||||||||||||||||||F||||||||||
OBX|1|ED|^Progress Notes|1|SEE ATTACHMENT||||||F|||
${ATTACH:d3ad-b33f-0000}
```

This is not valid HL7 and the parser barfs on it.

You can prove my idea by changing your incoming datatype to RAW. This will get past the parsing error (because it doesnt parse!) and you'll be able to see the incorrect HL7 structure.

Look at "Attachments Using Regex" at https://forums.mirthproject.io/articles/175046-using-attachments . You want a regex like the OBX regex shown in the example. That will make the large base64 encoded blob the attachment, not the whole segment, like this:

```
OBR|1|||||||^Progress Notes|||||||||||||||||||||F||||||||||
OBX|1|ED|^Progress Notes|1|SEE ATTACHMENT||||||F|||
PLY|||asdf|asdfs|${ATTACH:d3ad-b33f-0000}
```

**Reply by pacmano1** [reply]
What @jonbartels said.  You can delete the PLY segment though after the attachment is added using the regex,  It will be avaiable in the channel even with the PLY segment gone.

**Reply by pacmano1** [reply]
![image](https://github.com/nextgenhealthcare/connect/assets/44065187/5a6e8837-7460-4207-a107-53370c33eb1d)

`(?:PLY\|(?:[^|]*\|){3}(?:[^|^]*\^){0})([^|^\r\n]*)(?:[|^\r\n]|$)`

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/fb42e41e-ee71-4528-ac45-58ce68fb0a79)
